parl/remote/utils.py

```diff
@@ -52,49 +54,115 @@ def data_process():
+        class_name = cls.__name__
+
+        dumped_class_info = [
+            file_path, class_name, end_of_file, in_sys_path, sys.path
+        ]
     else:
-        module_name = module_name.lstrip('.' + os.sep).replace(os.sep, '.')
-        mod = __import__(module_name, globals(), locals(), [class_name], 0)
+        dumped_class_info = cls
+
+    return cloudpickle.dumps([defined_in_file, dumped_class_info])
 
-    cls = getattr(mod, class_name)
 
-    os.remove(tmp_file_name)
+def load_remote_class(remote_class_info):
+    """load a class given related info dumped in the client.
 
+    Args:
+      remote_class_info: [defined_in_file, dumped_class_info]
+                         - defined_in_file: whether the remote class is defined in a python file
+                         - dumped_class_info: information of dumped remote class (decided by `defined_in_file`).
+
+    Return:
+      cls: the class to load
+    """
+
+    defined_in_file, dumped_class_info = cloudpickle.loads(remote_class_info)
+
+    if defined_in_file:
+        """
+        dumped_class_info:
+            file_name: specify the file to load the class
+            class_name: specify the class to be loaded
+            end_of_file: line ID to indicate the last line that defines the class.
+            in_sys_path: whether the path of the remote class is in the environment path (sys.path).
+            client_sys_path: the environment paths of the client
+        """
+        file_name, class_name, end_of_file, in_sys_path, client_sys_path = dumped_class_info
+
+        with open(file_name + '.py') as t_file:
+            code = t_file.readlines()
+        code = simplify_code(code, end_of_file)
+        #folder/xx.py -> folder/xparl_xx.py
+        file_name = file_name.split(os.sep)
+        prefix = os.sep.join(file_name[:-1])
+        if prefix == "":
+            prefix = '.'
+
+        # Add pid to fix:
+        #    https://github.com/PaddlePaddle/PARL/issues/611
+        #    Multiple jobs may write the same file when `in_sys_path` is True.
+        new_file_name = 'xparl_{}'.format(os.getpid()) + file_name[-1]
+        module_name = prefix + os.sep + new_file_name
+        tmp_file_name = module_name + '.py'
+        with open(tmp_file_name, 'w') as t_file:
+            for line in code:
+                t_file.write(line)
+
+        if in_sys_path:
+            # append the environment paths of the client to the current environment path.
+            sys.path.extend(client_sys_path)
+
+            # the path of the remote class is in the sys.path, we can import it directly.
+            mod = __import__(new_file_name)
+        else:
+            module_name = module_name.lstrip('.' + os.sep).replace(os.sep, '.')
+            mod = __import__(module_name, globals(), locals(), [class_name], 0)
+
+        cls = getattr(mod, class_name)
+
+        os.remove(tmp_file_name)
+    else:
+        cls = dumped_class_info
```

Comment:
I'm afraid that the remote class distributed by cloudpickle supports the newest feature.
The users can access the attributes of remote instances, but it is not clear if another implementation supports this new feature.

Reply:
added unittest